File: day9/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func toIntList(txt string) ([]int, error) {
	vs := strings.Split(txt, " ")
	vsm := make([]int, len(vs))
	for i, v := range vs {
		n, err := strconv.Atoi(string(v))
		if err != nil {
			return nil, err
		}
		vsm[i] = n
	}
	return vsm, nil
}

func derive(line []int) []int {
	derivative := make([]int, len(line)-1)
	for c := 0; c < len(line)-1; c++ {
		derivative[c] = line[c+1] - line[c]
	}
	return derivative
}

func predict(line []int) (int, int) {
	if len(line) == 0 {
		return 0, 0
	}
	derivative := derive(line)
	d0, d1 := predict(derivative)
	return line[0] - d0, line[len(line)-1] + d1
}

func main() {

	file, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	sum1, sum2 := 0, 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		txt := scanner.Text()
		row, err := toIntList(txt)
		if err != nil {
			panic("integer expected")
		}
		prev, next := predict(row)
		fmt.Println(prev, ",", row, ",", next)
		sum1 += prev
		sum2 += next
	}

	err = scanner.Err()
	if err != nil {
		panic("input read error")
	}

	fmt.Println(sum1, sum2)
}

File: day9/main.py
s0, s1 = 0, 0
for line in open("day9/input.txt", "r"):
    nums = [int(p.strip()) for p in line.split(" ") if p.strip() != ""]
    derivatives = [nums]
    for _ in range(len(nums)):
        d = derivatives[-1]
        derivatives.append([d[i+1] - d[i] for i in range(len(d)-1)])
    for i in range(len(nums)-2, -1, -1):
        predictedNext = derivatives[i+1][-1] + derivatives[i][-1]
        predictedPrev = derivatives[i][0] - derivatives[i+1][0]
        derivatives[i] = [predictedPrev] + derivatives[i] + [predictedNext]
        print("   " * i, ("%5d " * len(derivatives[i])) % tuple(derivatives[i]))
    s0 += derivatives[0][0]
    s1 += derivatives[0][-1]
print(s0, s1)
